crates/bitwarden/src/state/state.rs

```diff
@@ -0,0 +1,293 @@
+            })
+            .await?;
+        drop(account);
+
+        // Save the new data
+        let profile = data.profile.ok_or(Error::MissingFields)?;
+
+        store_keys_from_sync(profile.as_ref(), self).await?;
+        store_profile_from_sync(profile.as_ref(), self).await?;
+
+        Ok(())
+    }
+
+    #[cfg(feature = "internal")]
+    pub(crate) async fn load_account(&self, id: Uuid) -> Result<()> {
+        let mut account = self.account.lock().await;
+        *account = StateStorage::new(id.to_string(), load_medium(&self.path));
+        account.read().await?;
+        Ok(())
+    }
+}
+
+pub struct StateStorage {
+    cache: HashMap<String, Value>,
+    account_id: String,
+    medium: Box<dyn StateStorageMedium>,
+}
+
+impl StateStorage {
+    fn new(account_id: String, medium: Box<dyn StateStorageMedium>) -> Self {
+        Self {
+            cache: HashMap::new(),
+            account_id,
+            medium,
+        }
+    }
+
+    pub fn get(&self) -> HashMap<String, Value> {
+        self.cache.clone()
+    }
+
+    #[cfg(feature = "internal")]
+    pub async fn read(&mut self) -> Result<HashMap<String, Value>> {
+        let value = self
+            .medium
+            .read()
+            .await?
+            .remove(&self.account_id)
+            .unwrap_or_default();
+        self.cache = value.clone();
+        Ok(value)
+    }
+
+    pub async fn modify<'b>(
+        &mut self,
+        modify_fn: impl FnOnce(&mut Option<HashMap<String, Value>>) -> Result<()> + Send + 'b,
+    ) -> Result<()> {
+        let account_id = self.account_id.clone();
+
+        let result = self
+            .medium
+            .modify(Box::new(move |state| {
+                let mut map = state.remove(&account_id);
+                modify_fn(&mut map)?;
+                if let Some(map) = map {
+                    state.insert(account_id, map);
+                }
+
+                Ok(())
+            }))
+            .await?;
+
+        self.cache = result.get(&self.account_id).cloned().unwrap_or_default();
+        Ok(())
+    }
+}
+
+fn load_medium(path: &Option<String>) -> Box<dyn StateStorageMedium> {
+    if let Some(path) = path {
+        #[cfg(target_arch = "wasm32")]
+        {
+            Box::new(WasmLocalStorageMedium::new(path.to_owned()))
+        }
+        #[cfg(not(target_arch = "wasm32"))]
+        {
+            Box::new(FileStateStorageMedium::new(path.into()))
+        }
+    } else {
+        Box::new(InMemoryStateStorageMedium::new())
+    }
+}
+
+type StateMap = HashMap<String, HashMap<String, Value>>;
+
+#[async_trait::async_trait]
+trait StateStorageMedium: Sync + Send + Debug {
+    async fn read(&mut self) -> Result<StateMap>;
+    async fn modify<'b>(
+        &mut self,
+        f: Box<dyn for<'a> FnOnce(&'a mut StateMap) -> Result<()> + Send + 'b>,
+    ) -> Result<StateMap>;
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+#[derive(Debug)]
+struct FileStateStorageMedium {
+    path: PathBuf,
+}
```

Comment:
We should probably move out wasm code to it's own file to easier conditionally compile it.